File: client/mobile/ios/App/App/MetorPushPlugin.swift
// The app's own push plugin (ADR-0017): asks for permission, registers with APNs, keeps the device's
// Web Push key pair, and reports the token and keys to the bridge (src/bridge.js), which hands the
// gateway a subscription whose endpoint is the relay. Registered in MainViewController.swift.
import Foundation
import Capacitor
import UserNotifications

@objc(MetorPushPlugin)
public class MetorPushPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "MetorPushPlugin"
    public let jsName = "MetorPush"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "register", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setComputer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearComputer", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setBadge", returnType: CAPPluginReturnPromise),
    ]
    static let approvalCategory = "metor.approval"   // notification actions Approve / Deny (registered in AppDelegate)
    static let opened = Notification.Name("metor.push.opened")
    static var pendingOpen: [String: String]?   // a tap (bot and computer) that arrived before the interface was listening
    private var pending: CAPPluginCall?

    public override func load() {
        let c = NotificationCenter.default
        c.addObserver(self, selector: #selector(gotToken(_:)), name: .capacitorDidRegisterForRemoteNotifications, object: nil)
        c.addObserver(self, selector: #selector(failed(_:)), name: .capacitorDidFailToRegisterForRemoteNotifications, object: nil)
        c.addObserver(self, selector: #selector(opened(_:)), name: MetorPushPlugin.opened, object: nil)
        if let open = MetorPushPlugin.pendingOpen { MetorPushPlugin.pendingOpen = nil; notifyListeners("opened", data: open, retainUntilConsumed: true) }
    }

    @objc func register(_ call: CAPPluginCall) {
        UNUserNotificationCenter.current().requestAuthorization(options: [.alert, .sound, .badge]) { granted, _ in
            guard granted else { call.resolve(["reason": "denied"]); return }
            self.pending = call
            DispatchQueue.main.async { UIApplication.shared.registerForRemoteNotifications() }
        }
    }
    // The bridge names the computer a push may come from; approvals are answered there straight from the notification
    @objc func setComputer(_ call: CAPPluginCall) {
        guard let id = call.getString("id"), let origin = call.getString("origin"), let token = call.getString("token") else { call.reject("id, origin and token"); return }
        do { try PushComputers.set(id, .init(origin: origin, token: token)); call.resolve() } catch { call.reject("keychain: \(error)") }
    }
    @objc func clearComputer(_ call: CAPPluginCall) {
        if let id = call.getString("id") { PushComputers.remove(id); PushBadges.remove(id); MetorPushPlugin.showBadge(PushBadges.total()) }
        call.resolve()
    }

    // The badge follows the bot list while the app is open: the unread total of that computer goes into the
    // per-computer store and the icon shows the sum over all computers; the notifications of bots that were
    // read disappear from the notification center (thread "bot:<computer>:<name>", the computer in metor_c)
    @objc func setBadge(_ call: CAPPluginCall) {
        let count = call.getInt("count") ?? 0
        let computer = call.getString("computer")
        let read = Set((call.getArray("read") as? [String]) ?? [])
        if let c = computer { PushBadges.set(c, count) }
        MetorPushPlugin.showBadge(computer == nil ? count : PushBadges.total())
        let center = UNUserNotificationCenter.current()
        if count == 0 && computer == nil { center.removeAllDeliveredNotifications() }
        else if count == 0 || !read.isEmpty {
            center.getDeliveredNotifications { delivered in
                let ids = delivered.filter { n in
                    let info = n.request.content.userInfo
                    if let c = computer, let from = info["metor_c"] as? String, from != c { return false }   // another computer's
                    if count == 0 { return true }
                    return read.contains((info["metor_bot"] as? String) ?? String(n.request.content.threadIdentifier.split(separator: ":").last ?? ""))
                }.map { $0.request.identifier }
                if !ids.isEmpty { center.removeDeliveredNotifications(withIdentifiers: ids) }
            }
        }
        call.resolve()
    }
    static func showBadge(_ count: Int) {
        DispatchQueue.main.async {
            if #available(iOS 16.0, *) { UNUserNotificationCenter.current().setBadgeCount(count) { _ in } } else { UIApplication.shared.applicationIconBadgeNumber = count }
        }
    }

    @objc func gotToken(_ n: Notification) {
        guard let token = n.object as? Data, let call = pending else { return }
        pending = nil
        do {
            let keys = try WebPushKeys.loadOrCreate()
            call.resolve(["platform": "ios", "token": token.hex, "sandbox": MetorPushPlugin.sandbox,
                          "p256dh": keys.privateKey.publicKey.x963Representation.base64url, "auth": keys.auth.base64url])
        } catch { call.reject("keys: \(error)") }
    }
    @objc func failed(_ n: Notification) {
        pending?.resolve(["reason": "apns: \((n.object as? Error)?.localizedDescription ?? "registration failed")"]); pending = nil
    }
    @objc func opened(_ n: Notification) {
        notifyListeners("opened", data: ["bot": (n.userInfo?["bot"] as? String) ?? "", "computer": (n.userInfo?["computer"] as? String) ?? ""], retainUntilConsumed: true)
    }
    // Development builds get sandbox tokens; App Store and TestFlight builds production ones
    static var sandbox: Bool {
        #if DEBUG
        return true
        #else
        return false
        #endif
    }
}

File: client/mobile/android/app/src/main/java/com/metor/mobile/MetorPushPlugin.java
// The app's own push plugin (ADR-0017): asks for the notification permission, fetches the FCM
// registration token, keeps the device's Web Push key pair, and reports token and keys to the bridge
// (src/bridge.js), which hands the gateway a subscription whose endpoint is the relay. A tap on a
// notification (MetorMessagingService) arrives here as the "opened" event. Registered in MainActivity.
package com.metor.mobile;

import android.content.Intent;
import android.os.Build;

import com.getcapacitor.JSObject;
import com.getcapacitor.PermissionState;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.google.firebase.FirebaseApp;
import com.google.firebase.messaging.FirebaseMessaging;

@CapacitorPlugin(name = "MetorPush", permissions = { @Permission(alias = "notifications", strings = { "android.permission.POST_NOTIFICATIONS" }) })
public class MetorPushPlugin extends Plugin {
    static final String EXTRA_BOT = "metor_bot";
    static final String EXTRA_COMPUTER = "metor_c";   // the computer a push came from (knowledge/design/several-computers.md)

    @Override
    public void load() { deliverOpened(getActivity().getIntent()); }

    @Override
    protected void handleOnNewIntent(Intent intent) { super.handleOnNewIntent(intent); deliverOpened(intent); }

    private void deliverOpened(Intent intent) {
        if (intent == null || !intent.hasExtra(EXTRA_BOT)) return;
        JSObject data = new JSObject(); data.put("bot", intent.getStringExtra(EXTRA_BOT)); data.put("computer", intent.getStringExtra(EXTRA_COMPUTER));
        intent.removeExtra(EXTRA_BOT); intent.removeExtra(EXTRA_COMPUTER);
        notifyListeners("opened", data, true);
    }

    // The bridge names the computer a push may come from; approvals are answered there straight from the notification
    @PluginMethod
    public void setComputer(PluginCall call) {
        String id = call.getString("id"), origin = call.getString("origin"), token = call.getString("token");
        if (id == null || origin == null || token == null) { call.reject("id, origin and token"); return; }
        try { WebPushCrypto.setComputer(getContext(), id, origin, token); call.resolve(); } catch (Exception e) { call.reject("store: " + e.getMessage()); }
    }
    @PluginMethod
    public void clearComputer(PluginCall call) { String id = call.getString("id"); if (id != null) WebPushCrypto.clearComputer(getContext(), id); call.resolve(); }

    // Android has no badge of its own: the launcher's dot or count comes from the active notifications. So the
    // notifications of bots that were read go away – of that computer (its id travels in the notification's
    // extras, MetorMessagingService), all of them when nothing is unread there any more. The count itself is
    // kept per computer for the number the next notification shows (the sum over all computers).
    @PluginMethod
    public void setBadge(PluginCall call) {
        android.app.NotificationManager nm = getContext().getSystemService(android.app.NotificationManager.class);
        int count = call.getInt("count", 0);
        String computer = call.getString("computer");
        if (computer != null) WebPushCrypto.setBadge(getContext(), computer, count);
        java.util.HashSet<String> read = new java.util.HashSet<>();
        com.getcapacitor.JSArray arr = call.getArray("read");
        if (arr != null) for (int i = 0; i < arr.length(); i++) { try { read.add(arr.getString(i)); } catch (Exception ignored) {} }
        if (count == 0 && computer == null) nm.cancelAll();
        else for (android.service.notification.StatusBarNotification n : nm.getActiveNotifications()) {
            android.os.Bundle x = n.getNotification().extras;
            String from = x == null ? null : x.getString(EXTRA_COMPUTER), bot = x == null ? null : x.getString(EXTRA_BOT);
            if (computer != null && from != null && !from.equals(computer)) continue;   // another computer's
            if (count == 0 || (bot != null && read.contains(bot))) nm.cancel(n.getId());
        }
        call.resolve();
    }

    @PluginMethod
    public void register(PluginCall call) {
        if (Build.VERSION.SDK_INT >= 33 && getPermissionState("notifications") != PermissionState.GRANTED) {
            requestPermissionForAlias("notifications", call, "permissionDone");
        } else {
            finishRegistration(call);
        }
    }

    @PermissionCallback
    private void permissionDone(PluginCall call) {
        if (getPermissionState("notifications") != PermissionState.GRANTED) { JSObject r = new JSObject(); r.put("reason", "denied"); call.resolve(r); return; }
        finishRegistration(call);
    }

    private void finishRegistration(PluginCall call) {
        if (FirebaseApp.getApps(getContext()).isEmpty()) { JSObject r = new JSObject(); r.put("reason", "fcm not configured (no google-services.json in this build)"); call.resolve(r); return; }
        FirebaseMessaging.getInstance().getToken().addOnCompleteListener(task -> {
            JSObject r = new JSObject();
            if (!task.isSuccessful() || task.getResult() == null) { r.put("reason", "fcm: " + (task.getException() != null ? task.getException().getMessage() : "no token")); call.resolve(r); return; }
            try {
                WebPushCrypto.Keys keys = WebPushCrypto.loadOrCreate(getContext());
                r.put("platform", "android"); r.put("token", task.getResult()); r.put("sandbox", false);
                r.put("p256dh", WebPushCrypto.base64url(keys.publicRaw())); r.put("auth", WebPushCrypto.base64url(keys.auth));
                call.resolve(r);
            } catch (Exception e) { call.reject("keys: " + e.getMessage()); }
        });
    }
}
